teps/0089-nonfalsifiable-provenance-support.md

```diff
@@ -0,0 +1,255 @@
+1. The Pod executes, and the entrypointer requests an SVID and signature over the Results
+1. Tekton Pipelines verifies the Results
+1. Meanwhile, Tekton Pipelines has been verifying that the TaskRun hasn't been modified during execution
+1. Tekton Pipelines requests a signature and SVID over the completed TaskRun yaml from SPIRE
+1. Tekton Pipelines stores the SVID and signature as annotations on the TaskRun
+1. Tekton Chains observes the TaskRun, and verifies that SVID and signature against the TaskRun yaml
+1. If verification is successful, Chains proceeds normally. Otherwise, it stops and doesn't sign anything!
+
+
+Then, Chains gets the TaskRun.
+Chains will:
+1. Verify the signature against the SPIRE SVID
+1. If verification fails, Chains will mark the TaskRun as "failed" and move on
+1. Otherwise, continue signing stuff! 
+
+## Implementation Plan
+The plan to achieve non-falsifiable provenance will be implemented in phases.
+
+**Phase 1**
+* Add support for Signed Results with SPIRE (this will primarily involve modifications to the entrypointer image)
+* Add support for Chains verifying Signed Results
+
+**Phase 2**
+* Implement Signed TaskRuns with SPIRE (requires further design)
+* Determine an alternate release process for the SPIRE feature (since the Pipelines Controller will need the SPIRE volume mounted in)
+* Add support for Chains verifying Signed TaskRuns
+
+Some ideas that have been mentioned for the release process:
+* Having a separate release yaml with the SPIRE volume mounted in (suggested by @bobcatwilson)
+* Using the Tekton operator to add the volume in (suggested by @vdemeester)
+
+In parallel with this work, we should:
+* Confirm that this meets the SLSA defintion of "non-falsifiable", which might require a security audit
+
+### Risks and Mitigations
+We'll need to depend on the [github.com/spiffe/go-spiffe](https://github.com/spiffe/go-spiffe) library to interact with the SPIRE agent, request SVIDs and signatures.
+
+
+## Design Details - Signed Results
+
+For now, this design section will focus on Part 1 of the implementation plan: Signed Results.
+This TEP will be updated as we flesh out the design for Signed Taskruns.
+
+### Enabling SPIRE on the Entrypointer Image
+We can add a feature flag `--enforce-nonfalsifiablity=spire` as described in [Customizing the Pipelines Controller behavior](https://github.com/tektoncd/pipeline/blob/main/docs/install.md#customizing-the-pipelines-controller-behavior) as an alpha feature.
+If the feature is enabled, then Pipelines would mount in the `csi` Volume into all Pods.
+
+The entrypointer image should also be able to see that this flag is set and accordingly sign Results.
+
+### Signing Results
+Once results are available to the entrypointer image, it will request a signature and SVID over each Results.
+The signature and SVID will be stored as annotations on the TaskRun, and can be verified by a client.
```

Comment:
is requesting the signature + svid for the results and adding them to the taskruns done at the end of the taskrun execution, or is it as each result is produced? asking because if it's the latter, I wonder how it'd work with long-running taskruns given that the signatures are short lived (and if it limits the timeouts)?

Reply:
we should add this while the results are added into taskRun status i.e. at the end of the taskRun execution. The `entrypoint` is aware of the result through the termination message but it does not classify as the `result` until taskRun finishes successfully. It is also possible that the task result is produced and part of the termination message but later if that taskRun fails, the result is lost and not available for consumption.

Reply:
Yup, it would be done at the end of taskrun execution, so the short lived signatures shouldn't be a problem here.

Reply:
@priyawadhwa just how short lived are the short lived signatures? im wondering b/c there could potentially be TaskRuns that take hours - maybe even days (or around a day?) - im not sure that we've ever set an upper limit on them (the longer the taskrun the more it would risk being rescheduled but that's another story 😅 )